config/store.go

```diff
@@ -52,15 +41,249 @@ type Store interface {
+	configNoEnv *model.Config
+
+	persistFeatureFlags bool
+}
+
+// Get fetches the current, cached configuration.
+func (s *Store) Get() *model.Config {
+	s.configLock.RLock()
+	defer s.configLock.RUnlock()
+	return s.config
+}
+
+// Get fetches the current, cached configuration without environment variable overrides.
+func (s *Store) GetNoEnv() *model.Config {
+	s.configLock.RLock()
+	defer s.configLock.RUnlock()
+	return s.configNoEnv
+}
+
+// GetEnvironmentOverrides fetches the configuration fields overridden by environment variables.
+func (s *Store) GetEnvironmentOverrides() map[string]interface{} {
+	return generateEnviromentMap(GetEnviroment())
+}
+
+// RemoveEnvironmentOverrides returns a new config without the environment
+// overrides
+func (s *Store) RemoveEnvironmentOverrides(cfg *model.Config) *model.Config {
+	s.configLock.RLock()
+	defer s.configLock.RUnlock()
+	return removeEnvOverrides(cfg, s.configNoEnv, s.GetEnvironmentOverrides())
+}
+
+// PersistFeatures sets if the store should persist feature flags.
+func (s *Store) PersistFeatures(persist bool) {
+	s.configLock.Lock()
+	defer s.configLock.Unlock()
+	s.persistFeatureFlags = persist
+}
+
+// Set replaces the current configuration in its entirety and updates the backing store.
+func (s *Store) Set(newCfg *model.Config) (*model.Config, error) {
+	s.configLock.Lock()
+	var unlockOnce sync.Once
+	defer unlockOnce.Do(s.configLock.Unlock)
+
+	oldCfg := s.config.Clone()
+
+	// Really just for some tests we need to set defaults here
+	newCfg.SetDefaults()
+
+	// Sometimes the config is received with "fake" data in sensitive fields. Apply the real
+	// data from the existing config as necessary.
+	desanitize(oldCfg, newCfg)
+
+	if err := newCfg.IsValid(); err != nil {
+		return nil, errors.Wrap(err, "new configuration is invalid")
+	}
+
+	newCfg = removeEnvOverrides(newCfg, s.configNoEnv, s.GetEnvironmentOverrides())
+
+	// Don't persist feature flags unless we are on MM cloud
+	// MM cloud uses config in the DB as a cache of the feature flag
+	// settings in case the managment system is down when a pod starts.
+	if !s.persistFeatureFlags {
+		newCfg.FeatureFlags = nil
+	}
+
+	if err := s.backingStore.Set(newCfg); err != nil {
+		return nil, errors.Wrap(err, "failed to persist")
+	}
+
+	if err := s.loadLockedWithOld(oldCfg, &unlockOnce); err != nil {
+		return nil, errors.Wrap(err, "failed to load on save")
+	}
+
+	return oldCfg, nil
+}
+
+func (s *Store) loadLockedWithOld(oldCfg *model.Config, unlockOnce *sync.Once) error {
+	configBytes, err := s.backingStore.Load()
+	if err != nil {
+		return err
+	}
+
+	var loadedConfig model.Config
+	if len(configBytes) != 0 {
+		if err = json.Unmarshal(configBytes, &loadedConfig); err != nil {
+			return jsonutils.HumanizeJsonError(err, configBytes)
+		}
+	}
+
+	loadedConfig.SetDefaults()
+
+	s.configNoEnv = loadedConfig.Clone()
+	fixConfig(s.configNoEnv)
+
+	loadedConfig = *applyEnviromentMap(&loadedConfig, GetEnviroment())
```

Comment:
Any reason to dereference it here?  We are taking the pointer again in the next few lines.

Reply:
Do you know if there are performance implications of this? Or just style?

Reply:
Mainly style. Basically it's weird to dereference something and then get a pointer to that again in multiple places. Since this function already returns a pointer, one would expect that to be used in future functions. And de-reference only when needed. Immediately de-referencing it means you have to take a pointer again, which is the majority of the next usage of the variable.

Reply:
Makes sense.